### PR TITLE
Several ops support zero dim on GPU and CPU

### PR DESCRIPTION
**Pull request by juncaipeng**
### PR types

Others
### PR changes

Others
### Describe

Several ops support zero dim on GPU and CPU
* paddle.nn.functional.softmax
* paddle.nn.functional.log_softmax
* paddle.nn.functional.gumbel_softmax
* paddle.nn.functional.prelu

### CODE REVIEW

File: python/paddle/fluid/tests/unittests/test_zero_dim_tensor.py [hunk 1]
```diff
@@ -1430,6 +1433,26 @@ def test_t(self):
...
+        self.assertEqual(out1.shape, [])
+        self.assertEqual(out1.grad.shape, [])
+        self.assertEqual(x.grad.shape, [])
+
+        w2 = paddle.full([], 0.25, dtype='float32')
+        out2 = paddle.nn.functional.prelu(x, w2)
+        out2.retain_grads()
+        out2.backward()
+        self.assertEqual(out2.shape, [])
+        self.assertEqual(out1.grad.shape, [])
```
Comment (zhwesky2010) [inline python/paddle/fluid/tests/unittests/test_zero_dim_tensor.py]: out2

File: python/paddle/fluid/tests/unittests/test_zero_dim_tensor.py [hunk 2]
```diff
@@ -84,6 +84,9 @@
     paddle.poisson,
     paddle.bernoulli,
     paddle.median,
+    paddle.nn.functional.softmax,
```
Comment (zhwesky2010) [inline python/paddle/fluid/tests/unittests/test_zero_dim_tensor.py:87]: 下个PR再加下test_zero_dim_tensor_xpu吧

File: paddle/phi/infermeta/binary.cc [hunk 3]
```diff
@@ -2104,7 +2104,7 @@ void PReluInferMeta(const MetaTensor& x,
                     MetaConfig config) {
   auto x_dim = x.dims();
   if (mode == "all") {
-    PADDLE_ENFORCE_EQ(phi::product(alpha.dims()),
+    PADDLE_ENFORCE_LE(phi::product(alpha.dims()),
```
Comment (zhwesky2010) [inline paddle/phi/infermeta/binary.cc]: 这里触发了报错吗，因为[]的product应该也是1，应该不会报错

File: python/paddle/nn/functional/activation.py [hunk 4]
```diff
@@ -495,12 +495,11 @@ def prelu(x, weight, data_format="NCHW", name=None):
             #    [ 6.  ,  7.  ,  8.  ,  9.  ]]]]
     """
     assert (
-        len(weight.shape) == 1
-    ), "The dim count of weight shape should be 1 in prelu()."
+        len(weight.shape) == 0 or len(weight.shape) == 1
```
Comment (zhwesky2010) [inline python/paddle/nn/functional/activation.py:498]: API文档里的 `The weight shape is [1] or [in]` 要改成 `The weight shape is [] or [in]`

File: paddle/phi/kernels/gpu/prelu_grad_kernel.cu [hunk 5]
```diff
@@ -77,10 +77,20 @@ class PreluOpGradFunctor {
...
     }
-    size_t plane_size = numel / input_dims[0] / input_dims[1];
-    size_t spatial_size = numel / input_dims[0];
-    size_t channel =
-        mode == ChannelLast ? input_dims[input_dims.size() - 1] : input_dims[1];
+
+    size_t plane_size;
+    size_t spatial_size;
+    size_t channel;
+    if (mode == PRELU_Scalar) {
```
Comment (zhwesky2010) [inline paddle/phi/kernels/gpu/prelu_grad_kernel.cu:84]: 这个PRELU_Scalar什么时候被设置的呀

Comment (juncaipeng) [reply paddle/phi/kernels/gpu/prelu_grad_kernel.cu:84]: 该文件底部  https://github.com/PaddlePaddle/Paddle/blob/c2ad8f4a181e8d556ce49e3e819eeec2c1290b51/paddle/phi/kernels/gpu/prelu_grad_kernel.cu#L154

File: python/paddle/fluid/tests/unittests/test_zero_dim_tensor.py (hunk 1 above)
Comment (juncaipeng) [reply python/paddle/fluid/tests/unittests/test_zero_dim_tensor.py]: done

File: python/paddle/nn/functional/activation.py (hunk 4 above)
Comment (juncaipeng) [reply python/paddle/nn/functional/activation.py:498]: done

File: paddle/phi/infermeta/binary.cc (hunk 3 above)
Comment (juncaipeng) [reply paddle/phi/infermeta/binary.cc]: 嗯  重测后确实是没报错了

File: python/paddle/fluid/tests/unittests/test_zero_dim_tensor.py [hunk 6]
```diff
@@ -1430,6 +1433,26 @@ def test_t(self):
...
+        out1.backward()
+        self.assertEqual(out1.shape, [])
+        self.assertEqual(out1.grad.shape, [])
+        self.assertEqual(x.grad.shape, [])
+
+        w2 = paddle.full([], 0.25, dtype='float32')
+        out2 = paddle.nn.functional.prelu(x, w2)
+        out2.retain_grads()
+        out2.backward()
+        self.assertEqual(out2.shape, [])
```
Comment (zhwesky2010) [inline python/paddle/fluid/tests/unittests/test_zero_dim_tensor.py:1452]: 加一下具体的前反向值的测试吧，可以下个PR加

**Review by zhwesky2010 — APPROVED** [review]

**Review by jzhang533 — APPROVED** [review]
